kernel/src/path.rs

```diff
@@ -1,136 +1,178 @@
-
-/// The delimiter to separate object namespaces, creating a directory structure. Note this is in url
-/// terms, so we use `/`
-const DELIMITER: &str = "/";
+use crate::{DeltaResult, Error, FileMeta, Version};
 
 /// How many characters a version tag has
 const VERSION_LEN: usize = 20;
 
 /// How many characters a part specifier on a multipart checkpoint has
 const MULTIPART_PART_LEN: usize = 10;
 
+/// The number of characters in the uuid part of a uuid checkpoint
+const UUID_PART_LEN: usize = 36;
+
 #[derive(Debug)]
-pub(crate) struct LogPath<'a> {
-    url: &'a Url,
-    pub(crate) filename: Option<&'a str>,
-    pub(crate) version: Option<Version>,
-    // if is compacted, this path spans version [`version`, `compacted_to_version`]
-    _compacted_to_version: Option<Version>,
-    pub(crate) is_commit: bool,
-    pub(crate) is_checkpoint: bool,
+#[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+#[cfg_attr(not(feature = "developer-visibility"), visibility::make(pub(crate)))]
+enum LogPathFileType {
+    Commit,
+    SinglePartCheckpoint,
+    #[allow(unused)]
+    UuidCheckpoint(String),
+    // NOTE: Delta spec doesn't actually say, but checkpoint part numbers are effectively 31-bit
+    // unsigned integers: Negative values are never allowed, but Java integer types are always
+    // signed. Approximate that as u32 here.
+    #[allow(unused)]
+    MultiPartCheckpoint {
+        part_num: u32,
+        num_parts: u32,
+    },
+    #[allow(unused)]
+    CompactedCommit {
+        hi: Version,
+    },
+    Unknown,
 }
 
-fn get_filename(path: &str) -> Option<&str> {
-    if path.is_empty() || path.ends_with('/') {
-        None
-    } else {
-        path.rsplit(DELIMITER).next()
+#[derive(Debug)]
+#[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+#[cfg_attr(not(feature = "developer-visibility"), visibility::make(pub(crate)))]
+struct ParsedLogPath<Location: CanTryIntoLogPath = FileMeta> {
+    pub location: Location,
+    pub filename: String,
+    pub extension: String,
+    pub version: Version,
+    pub file_type: LogPathFileType,
+}
+
+// Internal helper used by TryFrom<FileMeta> below. It parses a fixed-length string into the numeric
+// type expected by the caller. A wrong length produces an error, even if the parse succeeded.
+fn parse_path_part<T: FromStr>(
+    value: &str,
+    expect_len: usize,
+    location: &impl CanTryIntoLogPath,
+) -> DeltaResult<T> {
+    match value.parse() {
+        Ok(result) if value.len() == expect_len => Ok(result),
+        _ => Err(Error::invalid_log_path(location)),
     }
 }
 
-fn get_version_opt(version_str_opt: Option<&str>, expected_digits: usize) -> Option<Version> {
-    version_str_opt.and_then(|version_str| {
-        if version_str.len() == expected_digits {
-            version_str.parse().ok()
-        } else {
-            None
-        }
-    })
+// In prod, we ~always construct ParsedLogPath from FileMeta, but in testing it's convenient to use
```

Comment:
What's the opposite of "testing code"? "kernel" maybe? (also remove `~`)

```suggestion
// In kernel we always construct ParsedLogPath from FileMeta, but in testing it's convenient to use
```